File: core/src/main/java/no/elg/infiniteBootleg/world/ServerWorld.java
package no.elg.infiniteBootleg.world;

import static no.elg.infiniteBootleg.protobuf.Packets.DespawnEntity.DespawnReason.PLAYER_KICKED;
import static no.elg.infiniteBootleg.protobuf.Packets.DespawnEntity.DespawnReason.PLAYER_QUIT;

import com.badlogic.ashley.core.Entity;
import no.elg.infiniteBootleg.Main;
import no.elg.infiniteBootleg.protobuf.ProtoWorld;
import no.elg.infiniteBootleg.server.PacketExtraKt;
import no.elg.infiniteBootleg.world.ecs.components.required.PositionComponent;
import no.elg.infiniteBootleg.world.generator.ChunkGenerator;
import no.elg.infiniteBootleg.world.render.HeadlessWorldRenderer;
import org.jetbrains.annotations.NotNull;

/**
 * World with extra functionality to handle multiple players
 *
 * @author Elg
 */
public class ServerWorld extends World {

  private final HeadlessWorldRenderer render = new HeadlessWorldRenderer(this);

  public ServerWorld(@NotNull ProtoWorld.World protoWorld) {
    super(protoWorld);
  }

  public ServerWorld(@NotNull ChunkGenerator generator, long seed, @NotNull String worldName) {
    super(generator, seed, worldName);
  }

  public void disconnectPlayer(@NotNull String uuid, boolean kicked) {
    var player = getEntity(uuid);
    if (player != null) {
      removeEntity(player, kicked ? PLAYER_KICKED : PLAYER_QUIT);
    } else {
      Main.logger().warn("Failed to find player " + uuid + " to remove");
    }
  }

  @Override
  @NotNull
  public HeadlessWorldRenderer getRender() {
    return render;
  }

  @Override
  protected void syncAddEntity(@NotNull Entity entity, boolean loadChunk) {
    super.syncAddEntity(entity, loadChunk);
    //    if (entity instanceof Player player) {
    //      render.addClient(
    //        player.getUuid(),
    //        new ServerClientChunksInView(CoordUtil.worldToChunk(player.getPosition())));
    //    }
    render.update();

    Main.inst()
        .getScheduler()
        .executeSync(
            () ->
                PacketExtraKt.broadcastToInView(
                    PacketExtraKt.clientBoundSpawnEntity(entity),
                    entity.getComponent(PositionComponent.class).getBlockX(),
                    entity.getComponent(PositionComponent.class).getBlockY(),
                    null));
  }

  @Override
  public void removeEntity(@NotNull Entity entity) {
    super.removeEntity(entity);
    //    if (entity instanceof Player player) {
    //      render.removeClient(player.getUuid());
    //    }
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/world/ServerWorld.kt
package no.elg.infiniteBootleg.world

import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.EntityListener
import no.elg.infiniteBootleg.Main
import no.elg.infiniteBootleg.protobuf.Packets.DespawnEntity.DespawnReason
import no.elg.infiniteBootleg.protobuf.ProtoWorld
import no.elg.infiniteBootleg.server.broadcastToInView
import no.elg.infiniteBootleg.server.clientBoundSpawnEntity
import no.elg.infiniteBootleg.util.worldToChunk
import no.elg.infiniteBootleg.world.ecs.components.required.IdComponent.Companion.id
import no.elg.infiniteBootleg.world.ecs.components.required.PositionComponent
import no.elg.infiniteBootleg.world.ecs.components.required.PositionComponent.Companion.positionComponent
import no.elg.infiniteBootleg.world.ecs.playerFamily
import no.elg.infiniteBootleg.world.generator.ChunkGenerator
import no.elg.infiniteBootleg.world.render.HeadlessWorldRenderer
import no.elg.infiniteBootleg.world.render.ServerClientChunksInView

/**
 * World with extra functionality to handle multiple players
 *
 * @author Elg
 */
class ServerWorld : World {
  override val render = HeadlessWorldRenderer(this)

  constructor(protoWorld: ProtoWorld.World) : super(protoWorld)
  constructor(generator: ChunkGenerator, seed: Long, worldName: String) : super(generator, seed, worldName)

  fun disconnectPlayer(uuid: String, kicked: Boolean) {
    val player = getEntity(uuid)
    if (player != null) {
      removeEntity(player, if (kicked) DespawnReason.PLAYER_KICKED else DespawnReason.PLAYER_QUIT)
    } else {
      Main.logger().warn("Failed to find player $uuid to remove")
    }
  }

  override fun initializeEngine(): Engine = super.initializeEngine().also {
    it.addEntityListener(object : EntityListener {
      override fun entityAdded(entity: Entity): Unit = onEntityAdd(entity)
      override fun entityRemoved(entity: Entity): Unit = onEntityRemove(entity)
    })
  }

  private fun onEntityAdd(entity: Entity) {
    if (playerFamily.matches(entity)) {
      render.addClient(entity.id, ServerClientChunksInView(entity.positionComponent.x.worldToChunk(), entity.positionComponent.y.worldToChunk()))
    }
    render.update()
    Main.inst().scheduler.executeSync {
      broadcastToInView(
        clientBoundSpawnEntity(entity),
        entity.getComponent(PositionComponent::class.java).blockX,
        entity.getComponent(PositionComponent::class.java).blockY,
        null
      )
    }
  }

  private fun onEntityRemove(entity: Entity) {
    if (playerFamily.matches(entity)) {
      render.removeClient(entity.id)
    }
  }
}
